File: apps/api/src/services/analytics.service.ts
import type { EmailLogStatus, EmailTrackingEntry, PaginatedResult } from "@believe-ai/shared";
import { DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE } from "@believe-ai/shared";
import { analyticsRepository } from "../repositories/analytics.repository.js";
import { campaignRepository } from "../repositories/campaign.repository.js";
import { campaignService } from "./campaign.service.js";
import { emailLogRepository } from "../repositories/emailLog.repository.js";

interface EmailTrackingRow {
  _id: unknown;
  campaignId: unknown;
  contactId: unknown;
  userId: unknown;
  stepIndex: number;
  status: EmailLogStatus;
  providerMessageId: string | null;
  trackingToken: string;
  openCount: number;
  clickCount: number;
  replied: boolean;
  errorMessage: string | null;
  sentAt: Date | null;
  openedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
  campaignName: string;
  contactName: string;
  contactEmail: string;
}

function toEmailTrackingDto(row: EmailTrackingRow): EmailTrackingEntry {
  return {
    id: String(row._id),
    campaignId: String(row.campaignId),
    contactId: String(row.contactId),
    userId: String(row.userId),
    stepIndex: row.stepIndex,
    status: row.status,
    providerMessageId: row.providerMessageId,
    trackingToken: row.trackingToken,
    openCount: row.openCount,
    clickCount: row.clickCount,
    replied: row.replied,
    errorMessage: row.errorMessage,
    sentAt: row.sentAt ? new Date(row.sentAt).toISOString() : null,
    openedAt: row.openedAt ? new Date(row.openedAt).toISOString() : null,
    createdAt: new Date(row.createdAt).toISOString(),
    updatedAt: new Date(row.updatedAt).toISOString(),
    campaignName: row.campaignName,
    contactName: row.contactName,
    contactEmail: row.contactEmail,
  };
}

export const analyticsService = {
  async getDashboard(userId: string) {
    const [{ totalContacts, activeCampaigns, scheduledCampaigns, statusCounts }, recentCampaignDocs] =
      await Promise.all([analyticsRepository.dashboardCounts(userId), campaignRepository.list(userId)]);

    const counts: Partial<Record<EmailLogStatus, number>> = {};
    for (const row of statusCounts) counts[row._id as EmailLogStatus] = row.count;

    const emailsSent =
      (counts.SENT ?? 0) + (counts.DELIVERED ?? 0) + (counts.OPENED ?? 0) + (counts.CLICKED ?? 0) + (counts.REPLIED ?? 0);
    const opened = (counts.OPENED ?? 0) + (counts.CLICKED ?? 0) + (counts.REPLIED ?? 0);
    const clicked = (counts.CLICKED ?? 0) + (counts.REPLIED ?? 0);
    const replied = counts.REPLIED ?? 0;
    const rate = (n: number, d: number) => (d > 0 ? Math.round((n / d) * 1000) / 10 : 0);

    return {
      totalContacts,
      emailsSent,
      emailsDelivered: emailsSent,
      openRate: rate(opened, emailsSent),
      clickRate: rate(clicked, emailsSent),
      replyRate: rate(replied, emailsSent),
      activeCampaigns,
      scheduledCampaigns,
      recentCampaigns: recentCampaignDocs.slice(0, 5).map(campaignService.toDto),
    };
  },

  async getEmailTracking(userId: string, page?: number, limit?: number): Promise<PaginatedResult<EmailTrackingEntry>> {
    const pageNum = Math.max(1, page ?? 1);
    const limitNum = Math.min(MAX_PAGE_SIZE, Math.max(1, limit ?? DEFAULT_PAGE_SIZE));

    const { items, total } = await emailLogRepository.listByUserId(userId, pageNum, limitNum);

    return {
      items: (items as EmailTrackingRow[]).map(toEmailTrackingDto),
      page: pageNum,
      limit: limitNum,
      total,
      totalPages: Math.max(1, Math.ceil(total / limitNum)),
    };
  },
};

File: apps/backend/services/analytics_service.py
"""The outreach dashboard — mirrors apps/api's analytics.service.ts."""

from __future__ import annotations

import asyncio

from bson import ObjectId

from repositories import analytics_repository, email_log_repository
from schemas.analytics import DashboardStats, EmailTrackingEntry
from schemas.pagination import DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE, PaginatedResult, total_pages
from services.campaign_service import list_campaigns

_SENT_LIKE_STATUSES = ("SENT", "DELIVERED", "OPENED", "CLICKED", "REPLIED")
_OPENED_LIKE_STATUSES = ("OPENED", "CLICKED", "REPLIED")
_CLICKED_LIKE_STATUSES = ("CLICKED", "REPLIED")


def _rate(numerator: int, denominator: int) -> float:
    return round((numerator / denominator) * 1000) / 10 if denominator > 0 else 0.0


def _to_tracking_dto(row: dict) -> EmailTrackingEntry:
    return EmailTrackingEntry(
        id=str(row["_id"]),
        campaignId=str(row["campaignId"]),
        contactId=str(row["contactId"]),
        userId=str(row["userId"]),
        stepIndex=row["stepIndex"],
        status=row["status"],
        providerMessageId=row.get("providerMessageId"),
        trackingToken=row["trackingToken"],
        openCount=row["openCount"],
        clickCount=row["clickCount"],
        replied=row["replied"],
        errorMessage=row.get("errorMessage"),
        sentAt=row["sentAt"].isoformat() if row.get("sentAt") else None,
        openedAt=row["openedAt"].isoformat() if row.get("openedAt") else None,
        createdAt=row["createdAt"].isoformat(),
        updatedAt=row["updatedAt"].isoformat(),
        campaignName=row["campaignName"],
        contactName=row["contactName"],
        contactEmail=row["contactEmail"],
    )


async def get_dashboard(user_id: ObjectId) -> DashboardStats:
    (total_contacts, active_campaigns, scheduled_campaigns, counts), recent_campaigns = await asyncio.gather(
        analytics_repository.dashboard_counts(user_id), list_campaigns(user_id)
    )

    emails_sent = sum(counts.get(s, 0) for s in _SENT_LIKE_STATUSES)
    opened = sum(counts.get(s, 0) for s in _OPENED_LIKE_STATUSES)
    clicked = sum(counts.get(s, 0) for s in _CLICKED_LIKE_STATUSES)
    replied = counts.get("REPLIED", 0)

    return DashboardStats(
        totalContacts=total_contacts,
        emailsSent=emails_sent,
        emailsDelivered=emails_sent,
        openRate=_rate(opened, emails_sent),
        clickRate=_rate(clicked, emails_sent),
        replyRate=_rate(replied, emails_sent),
        activeCampaigns=active_campaigns,
        scheduledCampaigns=scheduled_campaigns,
        recentCampaigns=recent_campaigns[:5],
    )


async def get_email_tracking(user_id: ObjectId, page: int | None, limit: int | None) -> PaginatedResult[EmailTrackingEntry]:
    page_num = max(1, page or 1)
    limit_num = min(MAX_PAGE_SIZE, max(1, limit or DEFAULT_PAGE_SIZE))

    rows, total = await email_log_repository.list_by_user_id(user_id, page_num, limit_num)

    return PaginatedResult(
        items=[_to_tracking_dto(row) for row in rows],
        page=page_num,
        limit=limit_num,
        total=total,
        totalPages=total_pages(total, limit_num),
    )
